Web Inspector: Console: add internal properties for registrations of a `FinalizationRegistry`

dcrousso opened the pull request:
#### 626aa2207e3e6402daa4dfd62a02b6cf29ab8683
<pre>
Web Inspector: Console: add internal properties for registrations of a `FinalizationRegistry`
<a href="https://bugs.webkit.org/show_bug.cgi?id=276719">https://bugs.webkit.org/show_bug.cgi?id=276719</a>

Reviewed by Yusuke Suzuki.

Add special handling for `FinalizationRegistry` to make it easier for developers to understand its state (e.g. what is still alive vs what is dead but not yet notified vs etc.).

* Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
(Inspector::JSInjectedScriptHost::getInternalProperties):
Add internal properties for
- `cleanupCallback` to help the developer know what the `FinalizationRegistry` is for if the grab it via some general means (e.g. `queryInstances`).
- `live` is the list of objects that have not been GCd, each showing the `target`, `heldValue`, and (if provided) `unregisterToken`.
- `dead` is the list of metadata for objects that have been GCd (but not yet notified), each showing the `heldValue` and (if provided) `unregisterToken`.

* Source/JavaScriptCore/runtime/JSFinalizationRegistry.h:
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.cpp:
(JSC::JSFinalizationRegistry::forEachLive const): Added.
(JSC::JSFinalizationRegistry::forEachDead const): Added.
Expose ways for callers to get the above data.

* LayoutTests/inspector/model/remote-object/resources/remote-object-utilities.js:
* LayoutTests/inspector/model/remote-object/finalizationregistry.html: Added.
* LayoutTests/inspector/model/remote-object/finalizationregistry-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/282000@main">https://commits.webkit.org/282000@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a2c664f65aff8be64cd565e0a4b378445a2eb0d9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61698 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41052 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14290 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65677 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12243 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12514 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49767 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8506 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64767 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38142 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53449 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30599 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34801 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10679 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11174 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/54793 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56604 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10981 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67405 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60939 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5641 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10740 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57142 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5666 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53396 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57368 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13737 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4641 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82701 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14465 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37936 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39032 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37681 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
```diff
@@ -411,6 +412,50 @@ JSValue JSInjectedScriptHost::getInternalProperties(JSGlobalObject* globalObject
     }
 
+    if (JSFinalizationRegistry* finalizationRegistry = jsDynamicCast<JSFinalizationRegistry*>(value)) {
+        Locker locker { finalizationRegistry->cellLock() };
```
You cannot hold a lock while doing JS operations like putDirect etc. it can cause deadlock.

kmiller68 commented on Source/JavaScriptCore/runtime/JSFinalizationRegistry.h:
```diff
@@ -81,7 +81,10 @@ class JSFinalizationRegistry final : public JSInternalFieldObjectImpl<1> {
     void registerTarget(VM&, JSCell* target, JSValue holdings, JSValue token);
 
     JS_EXPORT_PRIVATE size_t liveCount(const Locker<JSCellLock>&);
+    void forEachLive(const Locker<JSCellLock>&, const Function<void(JSCell* target, JSValue heldValue, JSCell* unregisterToken)>&) const;
```
This could be a ScopedLambda or just lambda instead of a Function.

Constellation commented on Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
```diff
@@ -411,6 +412,57 @@ JSValue JSInjectedScriptHost::getInternalProperties(JSGlobalObject* globalObject
+            registrationObject->putDirect(vm, Identifier::fromString(vm, "heldValue"_s), liveRegistration.heldValue);
+            if (liveRegistration.unregisterToken)
+                registrationObject->putDirect(vm, Identifier::fromString(vm, "unregisterToken"_s), liveRegistration.unregisterToken);
+            liveArray->putDirectIndex(globalObject, liveIndex++, registrationObject);
```
This can throw an error, so error check is necessary.

Constellation commented on Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
```diff
@@ -411,6 +412,57 @@ JSValue JSInjectedScriptHost::getInternalProperties(JSGlobalObject* globalObject
+            registrationObject->putDirect(vm, Identifier::fromString(vm, "heldValue"_s), deadRegistration.heldValue);
+            if (deadRegistration.unregisterToken)
+                registrationObject->putDirect(vm, Identifier::fromString(vm, "unregisterToken"_s), deadRegistration.unregisterToken);
+            deadArray->putDirectIndex(globalObject, deadIndex++, registrationObject);
```
Ditto.

Constellation reviewed (approved):
r=me with comments.

dcrousso committed:
Web Inspector: Console: add internal properties for registrations of a `FinalizationRegistry`
https://bugs.webkit.org/show_bug.cgi?id=276719

Reviewed by Yusuke Suzuki.

Add special handling for `FinalizationRegistry` to make it easier for developers to understand its state (e.g. what is still alive vs what is dead but not yet notified vs etc.).

* Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
(Inspector::JSInjectedScriptHost::getInternalProperties):
Add internal properties for
- `cleanupCallback` to help the developer know what the `FinalizationRegistry` is for if the grab it via some general means (e.g. `queryInstances`).
- `live` is the list of objects that have not been GCd, each showing the `target`, `heldValue`, and (if provided) `unregisterToken`.
- `dead` is the list of metadata for objects that have been GCd (but not yet notified), each showing the `heldValue` and (if provided) `unregisterToken`.

* Source/JavaScriptCore/runtime/JSFinalizationRegistry.h:
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.cpp:
(JSC::JSFinalizationRegistry::forEachLive const): Added.
(JSC::JSFinalizationRegistry::forEachDead const): Added.
Expose ways for callers to get the above data.

* LayoutTests/inspector/model/remote-object/resources/remote-object-utilities.js:
* LayoutTests/inspector/model/remote-object/finalizationregistry.html: Added.
* LayoutTests/inspector/model/remote-object/finalizationregistry-expected.txt: Added.

Canonical link: https://commits.webkit.org/282000@main
